bindings/matlab/Visualization/mesh2tri/mesh2tri.m

```diff
@@ -0,0 +1,96 @@
+% introduced and therefore the original meshgrid formatted coordinates can
+% still be used for plotting, see examples below.
+% For cross division extra points are created at the centre of each
+% quadrilateral face using the mean of the input coordinates. The extra
+% coordinates are the last prod(size(X)-1) points (e.g.
+% V((numel(X)+1):end,:) )and can therefore be replaced by interpolated
+% coordinates if desired, see example.
+%
+%
+% %% EXAMPLE
+% clear all; close all; clc;
+%
+% [X,Y] = meshgrid(linspace(-10,10,25));
+% Z = sinc(sqrt((X/pi).^2+(Y/pi).^2));
+%
+% figure('units','normalized','Position',[0 0 1 1],'Color','w'); colordef('white');
+% subplot(2,2,1);
+% surf(X,Y,Z); hold on;
+% axis tight; axis square; grid on; axis off; view(3); view(-30,70);
+% title('Meshgrid','FontSize',20);
+%
+% [F,V]=mesh2tri(X,Y,Z,'f');
+% C=V(:,3); C=mean(C(F),2);
+% subplot(2,2,2);
+% patch('Faces',F,'Vertices',V,'FaceColor','flat','CData',C); hold on;
+% axis tight; axis square; grid on; axis off; view(3); view(-30,70);
+% title('Forward slash','FontSize',20);
+%
+% [F,V]=mesh2tri(X,Y,Z,'b');
+% C=V(:,3); C=mean(C(F),2);
+% subplot(2,2,3);
+% Example of using original meshgrid coordinates instead
+% trisurf(F,X,Y,Z);
+% axis tight; axis square; grid on; axis off; axis off; view(3); view(-30,70);
+% title('Back slash','FontSize',20);
+%
+% [F,V]=mesh2tri(X,Y,Z,'x');
+% Replace Z-coordinates of added points by interpolated values if desired
+% IND=(numel(X)+1):size(V,1);
+% ZI = interp2(X,Y,Z,V(IND,1),V(IND,2),'cubic');
+% V(IND,3)=ZI;
+%
+% C=V(:,3); C=mean(C(F),2);
+% subplot(2,2,4);
+% patch('Faces',F,'Vertices',V,'FaceColor','flat','CData',C); hold on;
+% axis tight; axis square; grid on; axis off; view(3); view(-30,70);
+% title('Crossed','FontSize',20);
+%
+%
+% Kevin Mattheus Moerman
+% kevinmoerman@hotmail.com
+% 15/07/2010
+%------------------------------------------------------------------------
+
+[J,I]=meshgrid(1:1:size(X,2)-1,1:1:size(X,1)-1);
+
+switch tri_type
+    case 'f'%Forward slash
+        TRI_I=[I(:),I(:)+1,I(:)+1;  I(:),I(:),I(:)+1];
+        TRI_J=[J(:),J(:)+1,J(:);   J(:),J(:)+1,J(:)+1];
+        F = sub2ind(size(X),TRI_I,TRI_J);
+    case 'b'%Back slash
+        TRI_I=[I(:),I(:)+1,I(:);  I(:)+1,I(:)+1,I(:)];
+        TRI_J=[J(:)+1,J(:),J(:);   J(:)+1,J(:),J(:)+1];
+        F = sub2ind(size(X),TRI_I,TRI_J);
+    case 'x'%Cross
+        TRI_I=[I(:)+1,I(:);  I(:)+1,I(:)+1;  I(:),I(:)+1;    I(:),I(:)];
+        TRI_J=[J(:),J(:);    J(:)+1,J(:);    J(:)+1,J(:)+1;  J(:),J(:)+1];
+        IND=((numel(X)+1):numel(X)+prod(size(X)-1))';
+        F = sub2ind(size(X),TRI_I,TRI_J);
+        F(:,3)=repmat(IND,[4,1]);
+        Fe_I=[I(:),I(:)+1,I(:)+1,I(:)]; Fe_J=[J(:),J(:),J(:)+1,J(:)+1];
+        Fe = sub2ind(size(X),Fe_I,Fe_J);
+        Xe=mean(X(Fe),2); Ye=mean(Y(Fe),2);  Ze=mean(Z(Fe),2);
+        X=[X(:);Xe(:)]; Y=[Y(:);Ye(:)]; Z=[Z(:);Ze(:)];
+end
+
+V=[X(:),Y(:),Z(:)];
+
+end

```

Comment:
Including a BSD-licensed file is ok, but for tracking correctly the vendored dependencies it is important: 
* To place them under the existing `extern` directory 
* Add a link from where did you get the files

Reply:
Copied what you did for the MoxUnit, so it will only be installed if enable Matlab or generate matlab. Is that ok?

Reply:
Can you please point to where the `IDYNTREE_INTERNAL_MESH2TRI_PATH` variable is used and where the `mesh2tri.m` file is installed? I was not able to find it.

Reply:
This is what I added in the extern folder:

```cpp
# mesh2tri ( https://it.mathworks.com/matlabcentral/fileexchange/28327-mesh2tri ) is a matlab function that allows converting a mesh format into a triangulation format.
# It is only installed if Matlab is enabled, not tested in Octave yet.
if (IDYNTREE_USES_MATLAB OR IDYNTREE_GENERATE_MATLAB) 
    # We save the location of mesh2tri in a variable that is then accessible to the test
    set(IDYNTREE_INTERNAL_MESH2TRI_PATH "${CMAKE_CURRENT_LIST_DIR}/mesh2tri/mesh2tri" PARENT_SCOPE)
endif()
```

But I think it is not correct. Could you point me into the right solution? I'm still not very proficient with CMakeLists...

Reply:
It is used to convert from the mesh type of varaible that matlab generates for the sphere and cylinder to a triangulation to keep the same format as the stl files and the box geometries.
Code line is here:
https://github.com/robotology/idyntree/blob/cfac9c4847855b85a7377bbab2b4afe771c66f24/bindings/matlab/%2BiDynTreeWrappers/getMeshes.m#L109

Reply:
I do not understand how this is working in your tests if you do not install `mesh2tri` . Where is this file supposed to be installed? I guess if you know its destination, you can just add a call to: 
~~~
install(FILES ${IDYNTREE_INTERNAL_MESH2TRI_PATH}/mesh2tri.m DESTINATION <whereYouWantToInstallIt>)
~~~

Reply:
For my tests, I was adding the path manually. I was testing mainly the functionality and not the install process. I will test the install process now.

Reply:
By default in many of the matlab repos that use bindings we add the install/mex folder so I will put the external visualization functions there directly. Should I put them inside a folder to signal they are for visualization or is fine just to add them there in the directory?